### ISSUE TITLE
arg_parser: bool values should have the default set to false

### ISSUE THREAD

**Issue by andreeaflorescu** [firecracker-microvm/firecracker#1685]
Right now the arg_parser only supports custom default values for `String` values. Specifying a custom default values for bool arguments (also called flags) doesn't make a lot of sense because when the flag is not passed as an argument it means that its value is false, and when the flag is passed as an argument its value is true. Even though custom default values don't make a lot of sense for `Value::bool` we should nevertheless have a default value for these types of arguments, and that default value should always be false.

I couldn't really find a clean way for implementing this, so ideas are welcomed.

One option that I was thinking about is to specify in the constructor of `Argument` the default value. Right now the constructor of `Argument` assumes that the type of the argument is flag (because it specifies that takes_value = false):
```rust
        Argument {
            name,
            required: false,
            requires: None,
            takes_value: false,
            default_value: None,
            help: None,
            user_value: None,
        }
```

Since this is the case, we could also set the default_value to Some(false).
When the `takes_value` function is called with `true`, we can reset the `default_value` to None in case that value was Some(false). This is a bit a not very nice workaround and is also a bit weird because the order in which you call `takes_value` and `default_value` matters.

Another option is to implement `default_value` for both `Value::String` and `Value::bool`, but for `Value::bool` only allow the default to be false.

From these two options, I think I prefer the second one because it is a less weird workaround :))

As I said both options are pretty meh. Other suggestions are welcomed.

**Comment by AnishShah** [comment]
I would like to take a stab at this.

**Comment by acatangiu** [comment]
One other solution would be:
```
--- a/src/utils/src/arg_parser.rs
+++ b/src/utils/src/arg_parser.rs
@@ -170,7 +170,7 @@ impl<'a> Argument<'a> {
 /// the argument takes a value, or `bool` if it's a flag.
 #[derive(Clone, Debug, PartialEq)]
 pub enum Value {
-    Bool(bool),
+    Bool,
     String(String),
 }
 
@@ -182,10 +182,10 @@ impl Value {
         }
     }
 
-    fn as_bool(&self) -> Option<bool> {
+    fn as_bool(&self) -> bool {
         match self {
-            Value::Bool(b) => Some(*b),
-            _ => None,
+            Value::Bool => true,
+            _ => false,
         }
     }
 }
@@ -222,11 +222,10 @@ impl<'a> Arguments<'a> {
             .and_then(|arg_value| arg_value.as_string())
     }
 
-    /// Return the value of an argument if the argument exists and has the type
-    /// bool. Otherwise return None.
-    pub fn value_as_bool(&self, arg_name: &'static str) -> Option<bool> {
+    /// Return whether the argument exists and has the type bool.
+    pub fn value_as_bool(&self, arg_name: &'static str) -> bool {
         self.value_of(arg_name)
-            .and_then(|arg_value| arg_value.as_bool())
+            .map_or(false, |arg_value| arg_value.as_bool())
     }
```

There's no point in having `Option<Bool(bool)>`. A param is either there, as `Some(Value::Bool)` or it's not.

**Comment by andreeaflorescu** [comment]
I really like the approach @acatangiu proposed.

I wouldn't change the enum variant from `Value::Bool(bool)` to `Value::Bool` even though it makes the code a bit simpler because it is also making it is a bit weird. We would end up with a `Value` type with no actual value.

**Comment by acatangiu** [comment]
The value would be encoded in the type 😄

**Comment by acatangiu** [comment]
Or instead of using a hierarchy of `Option<enum Value>` you could use flatten it in a single enum. Something like:
```rust
pub enum ArgValue {
    NotPresent,
    Present,
    PresentWithValue(String),
}

pub struct Argument<'a> {
    name: &'a str,
    required: bool,
    requires: Option<&'a str>,
    takes_value: bool,
    default_value: ArgValue,
    help: Option<&'a str>,
    user_value: ArgValue,
}
```

Which also simplifies things and is also _not weird_ like the previous suggestion 😛

**Comment by andreeaflorescu** [comment]
`Present` and `NotPresent` is the `Option<T>` in Rust so I don't think we should redefine that. I would stick with Option<T>. I also think that is makes sense for `value_as_bool` to still return an `None` for the case where the argument is invalid. 
Here is an use case where returning None for invalid flags makes sense:
- you want to see if the flag `--no-api` was passed.
- instead of writing `--no-api`, you write `--noapi` in the code by mistake.
- the behaviour I would expect is for the program to crash because I have a programming error in my code and I wrote `--noapi` instead of `--no-api`, but instead `value_as_bool` returns false. This wrong behaviour is not easy catch. Instead of my buggy program to crash fast, it misbehaves.

`Value::as_bool` can return just a bool because I don't think there is a better option to map the default value being false.

**Comment by lauralt** [comment]
I agree with @andreeaflorescu  that `values_as_bool` should still be able to return `None`, but I was thinking/hoping that we could still support this with the `ArgValue` enum.

**Comment by acatangiu** [comment]
> Present and NotPresent is the Option<T> in Rust so I don't think we should redefine that.

`Option` is an enum with two variants, our params have at least three states:
1. not present
2. present, takes no value
3. present, takes value

You could even extend it to define the value type at the same level:
```rust
enum ArgState {
   NotPreset,
   NoValue,
   Value(String),
   Value(usize),
   etc
}
```

`Option<enum X>` means a table of 2 x `sizeof X` states. The problem is some of the states modeled using the `Option<Value>` don't make sense like you rightly discovered and which led to opening this issue.

> instead of writing --no-api, you write --noapi in the code by mistake

Solving for this case using the `Option<Value>` is a mess for example 😄

**Comment by AnishShah** [comment]
@acatangiu I don't see how using `ArgState` or `ArgValue` in `Argument` struct is useful here. `default_value` will be `PresentWithValue`, `Present` or `NotPresent` depedending on `takes_value` field. This is a problem similar to what @andreeaflorescu mentioned in the [comment above](https://github.com/firecracker-microvm/firecracker/issues/1685#issue-584570240). It is weird because the order in which `takes_value` or `default_value` is called will matter.

I have another suggestion that is to explicitly differentiate between an Argument and a Flag - 

```
trait builder {
   fn default_value(.....) ....
   fn user_value(....) ...
   fn help( ...) ....
   fn required( ...) ....
   fn requires( ...) ....
}

impl builder for Argument {
}

impl builder for Flag {
}
````

Basically, we won't have to expose `takes_value` fn.

**Comment by andreeaflorescu** [comment]
@AnishShah would you then have initialize the `ArgParser` like this?

```rust
        ArgParser::new()
            .arg(
                Argument::new("exec-file")
                    .required(true)
                    .help("'exec-file' info."),
            )
            .arg(
                Flag::new("no-api")
                    .requires("config-file")
                    .help("'no-api' info."),
            )
```

That'd be pretty nit!

**Comment by AnishShah** [comment]
Yes. Should I go ahead and try implementing this design?

**Comment by andreeaflorescu** [comment]
I am super onboard and excited about this solution :laughing:. Let's see what @lauralt and @acatangiu say.

We probably need to do some renaming though. `Builder` is not saying a lot about the interface. We can't name it `Argument` because the implementation is `Argument`. Or we can name it `Argument` and then we can name the `Argument` implementation that takes values `ArgumentWithValue` and name the flags `Flag`? Other ideas are more than welcome.

**Comment by acatangiu** [comment]
Sounds good to me! 👍

**Comment by lauralt** [comment]
I like the idea too! Still thinking about other name for `ArgumentWithValue`, but no success so far. Maybe `NamedArgument`, but it's not very suggestive though.

**Comment by andreeaflorescu** [comment]
`NamedArgument` speaks more that the argument has a name, not that it has a value :)). Naming is hard.

**Comment by lauralt** [comment]
Yep, but apparently that is a word that is used for arguments that take values :sweat_smile:. But nevermind, it's not suggestive, I don't like it either.

**Comment by ioanachirca** [comment]
What do you think about naming the trait `Argument` and the 2 structs`ValueArgument` and `FlagArgument`? This might make the distinction between them clearer.

**Comment by AnishShah** [comment]
Thanks for the feedback. I'll send a new PR in few days since this involves a lot of changes.

**Comment by AnishShah** [comment]
Hi guys. When I implemented the above design, I realized it is not a good one because there was a lot of similar code in ValueArgument and FlagArgument struct impl. Instead I added two static methods - `value_argument` and `flag_argument` to Argument method in the new PR. Let me know your thoughts.

**Comment by alindima** [comment]
I think this is no longer applicable. We have reworked the arg_parser to have a flag type, which can be used like so: `arguments.flag_present("my-flag")`
